File: research/gaia-x/ir/runtime/src/process/operator/join/join.rs
//
//! Copyright 2021 Alibaba Group Holding Limited.
//!
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//!
//! http://www.apache.org/licenses/LICENSE-2.0
//!
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

use crate::error::FnGenResult;
use crate::process::functions::{JoinKeyGen, KeyFunction};
use crate::process::operator::keyed::KeySelector;
use crate::process::record::{Record, RecordKey};
use ir_common::generated::algebra as algebra_pb;
use ir_common::generated::algebra::join::JoinKind;

impl JoinKeyGen<Record, RecordKey, Record> for algebra_pb::Join {
    fn gen_left_key(&self) -> FnGenResult<Box<dyn KeyFunction<Record, RecordKey, Record>>> {
        Ok(Box::new(KeySelector::with(self.left_keys.clone())?))
    }

    fn gen_right_key(&self) -> FnGenResult<Box<dyn KeyFunction<Record, RecordKey, Record>>> {
        Ok(Box::new(KeySelector::with(self.right_keys.clone())?))
    }

    fn get_join_kind(&self) -> JoinKind {
        unsafe { ::std::mem::transmute(self.kind) }
    }
}

File: flex/engines/graph_db/runtime/adhoc/operators/join.cc
/** Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "flex/engines/graph_db/runtime/common/operators/join.h"
#include "flex/engines/graph_db/runtime/adhoc/operators/operators.h"

namespace gs {
namespace runtime {
bl::result<Context> eval_join(const physical::Join& opr, Context&& ctx,
                              Context&& ctx2) {
  JoinParams p;
  auto left_keys = opr.left_keys();
  for (int i = 0; i < left_keys.size(); i++) {
    if (!left_keys.Get(i).has_tag()) {
      LOG(ERROR) << "left_keys should have tag";
      RETURN_BAD_REQUEST_ERROR("left_keys should have tag");
    }
    p.left_columns.push_back(left_keys.Get(i).tag().id());
  }
  auto right_keys = opr.right_keys();
  for (int i = 0; i < right_keys.size(); i++) {
    if (!right_keys.Get(i).has_tag()) {
      LOG(ERROR) << "right_keys should have tag";
      RETURN_BAD_REQUEST_ERROR("right_keys should have tag");
    }
    p.right_columns.push_back(right_keys.Get(i).tag().id());
  }
  switch (opr.join_kind()) {
  case physical::Join_JoinKind::Join_JoinKind_INNER:
    p.join_type = JoinKind::kInnerJoin;
    break;
  case physical::Join_JoinKind::Join_JoinKind_SEMI:
    p.join_type = JoinKind::kSemiJoin;
    break;
  case physical::Join_JoinKind::Join_JoinKind_ANTI:
    p.join_type = JoinKind::kAntiJoin;
    break;
  case physical::Join_JoinKind::Join_JoinKind_LEFT_OUTER:
    p.join_type = JoinKind::kLeftOuterJoin;
    break;
  default:
    RETURN_UNSUPPORTED_ERROR("Unsupported join kind: " +
                             std::to_string(static_cast<int>(opr.join_kind())));
  }
  return Join::join(std::move(ctx), std::move(ctx2), p);
}
}  // namespace runtime
}  // namespace gs